hyperdome_client/server.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+"""
+Hyperdome
+
+Copyright (C) 2019 Skyelar Craver <scravers@protonmail.com>
+                   and Steven Pitts <makusu2@gmail.com>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+
+class Server:
+    """
+    Holder class for server connection details
+    """
+
+    class InvalidOnionAddress(Exception):
+        """
+        The onion address provided does not contain a valid v3 public key
+        """
+
+        pass
+
+    def __init__(self, url="", nick="", uname="", passwd="", is_counselor=False):
+        self.url = url.strip()
+        if url:
+            self._check_url()
+        self.nick = nick
+        self.username = uname
+        self.password = passwd
+        self.is_counselor = is_counselor
+
+    def _check_url(self):
+        """
+        Ensure URL is properly formatted
+        """
+
+        if not (self.url.startswith(("http://", "https://"))):
+            self.url = f"http://{self.url}"
+        if not self.url.endswith(".onion"):
+            self.url = f"{self.url}.onion"
+
+        onion_key = self.url[7:-6]
```

Comment:
Might mess up for https, no?

Reply:
onions don't support https because SSL requires a certificate authority which verifies the identity of the host, which removes anonymity for the hidden service. Instead the onion v3 protocol does its own fancy end to end encryption that verifies the identity of the server with its own public key.

Reply:
Shouldn't you remove https from the earlier check, then? (If startswith http or https)?

Reply:
I was under the impression that it was valid to use https, because in the tor browser you can use https to access .onion URLs just fine, but after testing I realize that actually trying to load an onion with https fails, so yes it should be removed in the above / replaced with http if https is given. Maybe It should just strip up to // if it exists and replace with http?

I'll make note and fix it later